Po certchain recv

arturkow2000 opened the pull request.

arturkow2000 committed:
Support embedding certificate

Fobnail requires Platform Owner root certificate embedded into its
firmware. From now, during build, pem certificate is read from file,
converted to DER and automatically embedded.

arturkow2000 committed:
certmgr: support veryfing of PO certificates

This commit introduces two verification modes, one is for verifying EK
certificates and it retains previous behaviour, new mode is for PO
(Platform Owner) certificates and it has applies different verification
policy. Additionally, from now on, certificates may be temporarily
injected (are gone with certmgr instance), this is used for PO chain
verification.

arturkow2000 committed:
certmgr: make X509Certificate Cloneable, add certificate_raw method

certificate_raw method returns X.509 certificate as DER, this will be
used for comparing loaded certificates for equality.

arturkow2000 committed:
Begin rework certificate verification

We no longer allow trusted certificates to be installed to LittleFS -
from now on, all trusted certificates are embedded into firmware during
build and cannot be replaced except by firmware update. Directory
structure inside certstore has been flattened to reduce overhead with
LittleFS reserving 8 KiB of storage (2 blocks) for each new directory.
Add support for installing certchain (few certificates bundled into a
single file), this is currently usd for storing PO certchain in a single
file with well-known name.

arturkow2000 committed:
Implement PO certchain receiving, verification and saving to storage

arturkow2000 committed:
certmgr: cleanup unneeded code, unify database view

Remove code that allowed to verify certificates without Authority Key
ID. TCG Credential Specification requires this extensions to be present
on all certificates forming EK cert chain. Unify certificate DB view,
provided method will load certificates from any source (embedded,
volatile, filesystem), this greatly reduces complexity.

arturkow2000 committed:
Don't save PO cert chain

There is no need to save PO certchain as we use it only during token
provisioning.

krystian-hebel commented on src/certmgr/store.rs:
```diff
@@ -1,170 +1,84 @@
-use super::{CertMgr, X509Certificate};
+use alloc::vec::Vec;
+use serde::Serialize;
+use trussed::types::{Location, Message, PathBuf};
```
Unused imports (`Vec`, `Serialize` and `Message`)

arturkow2000 replied:
Fixed

krystian-hebel commented on src/client/token_provisioning/mod.rs:
```diff
@@ -144,6 +164,81 @@ impl<'a> FobnailClient<'a> {
+        }
+    }
+
+    /// Verify certchain and save it to persistent storage if verification is
```
Drop the part about saving.

arturkow2000 replied:
Done

krystian-hebel commented on src/client/token_provisioning/mod.rs:
```diff
@@ -144,6 +164,81 @@ impl<'a> FobnailClient<'a> {
+            }
+        }
+
+        certmgr.clear_volatile_certs();
```
Assuming we want to check newly created certificate in 4.5, can we keep volatile certs until then?

arturkow2000 replied:
We can.
Certs are now kept till the end of provisioning, and are cleared from SignalStatus both on error and success.

arturkow2000 committed:
build.rs: rebuild if PO root certificate has changed

arturkow2000 committed:
Clear volatile before retrying token provisioning

Move volatile cert clearing into SignalStatus so that certs are always
cleared on error.

arturkow2000 committed:
Fix broken error handling in verify_certchain

krystian-hebel commented on src/client/token_provisioning/mod.rs:
```diff
@@ -144,6 +169,78 @@ impl<'a> FobnailClient<'a> {
+    }
+
+    /// Verify certchain. On success certificates are loaded into certstore as
+    /// volatile certificates so that they can be used during later.
```
```suggestion
    /// volatile certificates so that they can be used later.
```

arturkow2000 replied:
Fixed

krystian-hebel reviewed (commented):
Volatile certificates should probably be flushed in state that verifies new certificate sent by PO, but since it doesn't exist yet it can happen in SignalStatus.

CI fails due to lack of `root.crt`, not sure if we want to fix it now or not.

arturkow2000 commented:
@krystian-hebel I'm clearing certificates from SignalStatus because it's common path both on error and success. Otherwise we would have to call `clear_volatile_certs` before each call to `state.error()`.

arturkow2000 committed:
build.rs: add FOBNAIL_PO_ROOT variable

FOBNAIL_PO_ROOT must point to PEM certificate which will be installed as
Platform Owner root certificate.

arturkow2000 committed:
Fix CI

Fobnail now requires Platform Owner certificate to build - add dummy
certificate to CI scripts

arturkow2000 committed:
Install dummy certificate to source directory instead of temp directory

Docker mounts source directory only.

arturkow2000 committed:
Move hardcoded certificate into separate file

This prevents typos hook from complaining.

krystian-hebel commented on .github/workflows/dummy_po_root.crt:
```diff
@@ -0,0 +1,22 @@
+-----BEGIN CERTIFICATE-----
```
We probably should change issuer/subject names so that they say that this certificate should not be used in production systems, in case somebody tries to use artifacts produced by CI.

arturkow2000 replied:
Done

krystian-hebel commented on src/client/token_provisioning/mod.rs:
```diff
@@ -144,6 +169,78 @@ impl<'a> FobnailClient<'a> {
+        /// Minimal number of certificates required in a chain (including root).
+        const MIN_CERTS: usize = 2;
+        /// Max number of certificates allowed in a chain (including root).
+        const MAX_CERTS: usize = 3;
```
If we want to use the same function later for checking whether newly created Fobnail certificate is valid, this should be 4.

arturkow2000 replied:
@krystian-hebel We won't use the same function, PO chain is still loaded and we only need to call `certmgr.verify()`

arturkow2000 committed:
dummy_po_root.crt: this for development use only

Update with a certificate that clearly states this is for development
use only.
